File: python/143. Reorder List.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if not head or not head.next: return
        p = head
        st = []
        while p:
            st.append(p)
            p = p.next
        
        dummy = ListNode(0)
        p = dummy
        while len(st)>1:
            p.next = st.pop(0)
            p = p.next
            print p.val
            p.next = st.pop()
            p = p.next
            print p.val

        if st:
            p.next = st[-1]
            p = p.next
        
        p.next = None

            
if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    su = Solution()
    su.reorderList(head)                            

File: c++/143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head) return;
        
        ListNode *p = head;
        int len = 0, cnt = 0;
        while (p) {
            len++;
            p = p->next;
        }
        
        p = head;
        while (p->next && cnt < len/2 - 1) {
            cnt++;
            p = p->next;
        }
        
        ListNode* new_head = new ListNode(0), *p1, *t, *q, *r;
        p1 = p->next;
        p->next = NULL;
        while (p1) {
            t = p1->next;
            p1->next = new_head->next;
            new_head->next = p1;
            p1 = t;
        }
        
        //merge
        p = head; q = new_head->next; r = new_head;
        while (p && q) {
            r->next = p;
            t = p->next;
            p->next = q;
            p = t;
            r = q;
            q = q->next;
        }
        
        head = new_head->next;
        delete(new_head);
    }
};